[g4star] Miscellaneous updates to StarAgmlLib

klendathu2k opened the pull request:
Merging StarAgmlLib which are independent of changes to
the StGeant4Maker development.

Brings code up to commit
71921c98f22fd5f8378ee394b1e546875743a3b2

The changes represented in the next commit appears to have already been integrated.
7dcbd58301d1e9a4eb9611256491c5b981069070

Beyond that commit (starting with 4c5edb1d9891627c6f0f56e6a1d1d4f52cc27646),
changes to StarAgmlLib are coupled to changes introduced in the StGeant4Maker/
directory.

- Agml modules extended with hit scoring

klendathu2k committed:
Add map of GEANT3 paramaters and cuts associated to volumes.

klendathu2k committed:
AgMLExtension inherits from TGeoExtension, which allows user and/or
framework information to be attached to TGeoVolumes.  The VMC framework
utlizes the "framework" extensions, so we will utilize the "user"
extensions in AgML.

- Attach user-defined hit scoring functions to sensitive volumes.
  When the MC application's hit scoring is invoked in a given
  volume, the general hit scoring function can invoke any user
  functions stored on the volume.

- Store the number of volume branchings for each volume, allowing
  us to define a unique volume ID to each volume defined within
  an AgML module.

klendathu2k committed:
Cleanup memory.

klendathu2k committed:
General cleanup, and moving mixture component to protected.

klendathu2k committed:
AgML: Introduce a namespace, and uncomment a few trig functions

AgMath provides math functions used by AgML codes. A few of these are
functions defined in FORtran. A few defines have been made to prevent
AgML variables from being mistaken for c-math routines.

klendathu2k committed:
Attach AgMLExtensions to the modules

klendathu2k committed:
Cleanup and replace ROOT types with basic types

klendathu2k committed:
WARN: Disables misalignment when run under ROOT6.  Still
enabled under ROOT5.

klendathu2k committed:
We don't even need a classdef here...

klendathu2k committed:
We don't need to access most of this interactively... so comment out pragmas...

klendathu2k committed:
Responsible for setting up the AgMLExtensions in the TGeoVolumes,
which connect user codes and volume IDs from the AgML source code
to the TGeo geometry model.

klendathu2k committed:
Back to standard StMessMgr... ENDL-->endm

klendathu2k committed:
Back to standard StMessMgr... ENDL-->endm

klendathu2k committed:
The sanity check on the main branch is newer than the geant4star version.

klendathu2k committed:
Expose pi to geoemtry modules

plexoos commented on StarVMC/StarAgmlLib/StarAgmlLibLinkDef.h:
```diff
@@ -3,24 +3,18 @@
+//#pragma link C++ class AgPosition+;
+//#pragma link C++ class AgShape+;
+//#pragma link C++ class AgStructure+;
+//#pragma link C++ class AgVolume+;
```
```suggestion
```

klendathu2k replied:
Ok.

plexoos commented on StarVMC/StarAgmlLib/AgMLStructure.h:
```diff
@@ -220,7 +220,12 @@ class AgMLStructure : public AgMLStructureBase
-
+ ~AgMLStructure() {   
+   for ( auto t : table ) { // cleanup the table
+     if (t) delete (t);
```
```suggestion
     delete t;
```

delete works for nullptr's

klendathu2k replied:
I may be over / underthinking the problem here... but are we guaranteed that we won't recieve a zero from TTable?

plexoos commented on StarVMC/StarAgmlLib/AgMath.h:
```diff
@@ -67,10 +69,10 @@ template <typename T> T min(const T &a, const T &b ){ return TMath::Min(a,b); }
-// #define sin(x) TMath::Sin( (Double_t) x )
-// #define tan(x) TMath::Tan( (Double_t) x )
-// #define exp(x) TMath::Exp( (Double_t) x )
+#define cos(x) TMath::Cos( (Double_t) x )
+#define sin(x) TMath::Sin( (Double_t) x )
+#define tan(x) TMath::Tan( (Double_t) x )
+#define exp(x) TMath::Exp( (Double_t) x )
```
Are these defines really necessary? Don't they collide with the ones from the standard library?

klendathu2k replied:
I believe that I put those in place to minimize numerical differences between the ROOT and GEANT positions and / or dimensions of a fairly small number of volumes.

klendathu2k replied:
Okay... a little archeology is important here.

From commit fd6f8ab786fa7bf435ba290e9059556f7c56bc24

    sin, cos, tan, and exp need to be computed at highest precision possible in order to minimize
    introduction of roundoff error in results.

There seems to be a history of people reflexively removing this code...
fc2109800a5a00e3c915b5b1c8952a8646d38d4c

... and I keep having to restore it...
202690286974776a0098f4f2ae2368c8e5e7c407

So, yes.  These definitions are considered to be necessary.  The goal being to minimize roundoff errors between the geant3 and root geometry models.  Removal would minimally require material budget-level consistency checks to be made to validate geometries.

It is worth adding some comments for posterity into AgMath.h methinks.

plexoos replied:
Fine as long as it works...

plexoos commented on StarVMC/StarAgmlLib/AgPlacement.h:
```diff
@@ -36,63 +36,53 @@ class AgPlacement : public TNamed, public AgParameterList<double>
-  std::vector< Int_t    > mType;		//!
+#if  !defined(__CLING__)
+  std::vector< Rotation > mRotation;
+#endif
```
What is the reason to hide this from rootcling? Did you mean rootcint?

klendathu2k replied:
No, I did mean cling here.  I have a vague recollection that cling was giving me an error on that particular line of code, and so I hid it.

plexoos replied:
How about we remove these guards for now and see if the compilation fails with ROOT6. If it does, adding a commit referencing the error message will be most helpful

plexoos replied:
Addressed in 1bc5b0c4fd923ab1fdcd2c55ad600c42ae411f28

plexoos commented on StarVMC/StarAgmlLib/Mortran.h:
```diff
@@ -64,7 +64,7 @@ template <typename T> class Array_t
   Int_t _min, _max;
 
-  ClassDef(Array_t,0);
+  //c l a s s d e f (Array_t,0);
```
```suggestion
```

plexoos commented on StarVMC/StarAgmlLib/AgPosition.cxx:
```diff
@@ -35,7 +36,12 @@ void AgMLDbMatrix::operator()( const char* table, const int row, double matrix[4
+  TTable* ttable = (TTable *)StMaker::GetChain() -> GetDataBase( tablename );
+#else
+  TTable* ttable = 0;
+#endif
```
This is probably the most controversial change in this PR since you deliberately break the code for ROOT6. If we leave it as is it maybe easier to see the problem when use it with ROOT6

klendathu2k replied:
I believe I did this anticipating problems with GetDataBase( tablename) under root 6.  

GetDataBase loads / unloads macros which look like this:
https://github.com/star-bnl/star-sw/blob/main/StarDb/Geometry/pxl/pxlLadderOnSectorMisalign.20131210.000001.C
https://github.com/star-bnl/star-sw/blob/main/StarDb/Geometry/ist/istLadderOnIstMisalign.20131210.000001.C

They all have the same function defined...  CINT handles this reasonably well, as the interpreter can just drop the function.
When I started geant4star, it was an open question whether or not ROOT6 would handle this gracefully.  Once we are ready to use ROOT 6 in production, we can restore this line.

plexoos replied:
Addressed in 1bc5b0c4fd923ab1fdcd2c55ad600c42ae411f28

plexoos reviewed (commented):
Unless I am missing something here, I think this PR can be merged as it appears not to break the existing chains.

klendathu2k commented:
... will take a stab at requested changes shortly...

klendathu2k committed:
Cleanup linkdef

klendathu2k committed:
There is a pretty strong guarentee that table entries are not null
here b/c AgML would not be filling / creating null entries.  Plus...
it should be safe to delete a nullptr...

klendathu2k committed:
Cautionary comments for future code maintainers...

klendathu2k committed:
Rid ourselves of that commented-out classdef...

plexoos commented on StarVMC/StarAgmlLib/AgBlock.cxx:
```diff
@@ -5,12 +5,13 @@ ClassImp(AgBlock);
 #include "StarAgmlStacker.h"
-#include "StMessMgr.h"
+#include "StMessMgr.h"                                                                                                                  
+#include <iostream>
```
This file is already included on line 5. Are you actually using anything from this header in this file? If not, please remove.

plexoos replied:
```suggestion
```

plexoos commented on StarVMC/StarAgmlLib/AgMLExtension.cxx:
```diff
@@ -1,6 +1,15 @@
+				 mSensitive(0), 
+				 mTracking(0),
+				 mBranchings(0),
+				 mVolumeId( new AgMLVolumeId ),
```
You may have a memory leak here. It does not look like you release the resource in the implementation file

plexoos replied:
Resolved by 3c7beb3161ba747b14a1dc4d89e6d02fb5d37f3d

plexoos commented on StarVMC/StarAgmlLib/AgMLExtension.h:
```diff
@@ -15,11 +30,36 @@ class AgMLExtension : public TGeoRCExtension {
+  void SetBranchings( int b ) { mBranchings=b; }  
+
+  void SetVolumeIdentifier( AgMLVolumeId* identifier ){ mVolumeId = identifier; }
+  void AddHitScoring( AgMLScoring* sc ){ mHitScoring.push_back( sc ); }
```
It is not obvious who owns the objects passed by pointer. If it is the mHitScoring container then you need to delete them in this class

plexoos commented on StarVMC/StarAgmlLib/AgMath.cxx:
```diff
@@ -1,5 +1,8 @@
 #include "AgMath.h"
 #include "TMath.h"
+//using namespace AgMath;
```
```suggestion
```

plexoos commented on StarVMC/StarAgmlLib/AgPlacement.h:
```diff
@@ -36,63 +36,53 @@ class AgPlacement : public TNamed, public AgParameterList<double>
   enum { kUnknown, kRot3, kRot6, kRotO };
   enum { kAlphaX,  kAlphaY, kAlphaZ };
 
+  //#ifndef __CINT__
```
```suggestion
```

plexoos commented on StarVMC/StarAgmlLib/AgModule.h:
```diff
@@ -5,6 +5,7 @@
 #include "Mortran.h"
 #include "AgDetp.h"
 #include <string>
+#include <StarVMC/StarAgmlLib/AgMLExtension.h>
```
```suggestion
#include "StarVMC/StarAgmlLib/AgMLExtension.h"
```

plexoos commented on StarVMC/StarAgmlLib/AgPosition.cxx:
```diff
@@ -4,13 +4,15 @@
-#include <StMessMgr.h>
+#include "RVersion.h" 
+#include "StMessMgr.h"                                                                                                                  
+#include <StChain.h>
```
```suggestion
#include "StChain.h"
```

plexoos commented on StarVMC/StarAgmlLib/StarAgmlLibLinkDef.h:
```diff
@@ -1,26 +1,10 @@
+#pragma link C++ class AgMLExtension+;
 #endif
+
+
```
```suggestion
```

plexoos commented on StarVMC/StarAgmlLib/StarTGeoStacker.cxx:
```diff
@@ -639,12 +638,25 @@ Bool_t StarTGeoStacker::Build( AgBlock *block )
       // AgML extented volume information
       AgMLExtension* agmlExt = gAgMLExt[ block->GetName() ]; 
-      if ( 0==agmlExt ) { gAgMLExt[ block->GetName() ] = agmlExt = new AgMLExtension(); }
+      if ( 0==agmlExt ) { 
```
```suggestion
      if ( !agmlExt ) { 
```

plexoos commented on StarVMC/StarAgmlLib/StarTGeoStacker.cxx:
```diff
@@ -720,12 +732,25 @@ Bool_t StarTGeoStacker::Build( AgBlock *block )
 
 	  AgMLExtension* agmlExt = gAgMLExt[ block->GetName() ]; 
-	  if ( 0==agmlExt ) { gAgMLExt[ block->GetName() ] = agmlExt = new AgMLExtension(); }
+	  if ( 0==agmlExt ) { 
```
```suggestion
	  if ( !agmlExt ) { 
```

plexoos commented on StarVMC/StarAgmlLib/AgMath.h:
```diff
@@ -1,12 +1,24 @@
 // Provides single point to define and/or redefine mathematical
 // keywords used in AgSTAR syntax.  i.e. cos, sin, cosh, tanh, ...
+//
+// Each of the definitions below were added for added to either workaround
```
```suggestion
// Each of the definitions below were added to either workaround
```

plexoos commented on StarVMC/StarAgmlLib/AgMath.h:
```diff
@@ -86,7 +97,11 @@ class AgPower
+
 //Float_t operator^(const Float_t &a, const Float_t &b);
 using namespace std;
+using namespace AgMath;
```
Doing this in a header file is of course a very bad practice, especially with the std namespace. I assume this header will never be public

plexoos committed:
Update StarVMC/StarAgmlLib/AgMath.h

plexoos committed:
Apply suggestions from code review

plexoos committed:
Fix memory leak in AgMLExtension

plexoos committed:
Let's confirm the problem with ROOT6 first and then fix

https://github.com/star-bnl/star-sw/pull/308/files#r882223665

plexoos commented:
There are no unresolved conversations and all checks have passed. We can merge if there are no further comments.

plexoos commented:
@klendathu2k OK to merge?

plexoos commented:
Ping @klendathu2k

klendathu2k commented:
> @klendathu2k OK to merge?

Yes, good to merge.

klendathu2k commented:
Yes.  Merge away.

On 2022-08-25 11:04, Dmitri Smirnov wrote:
> @klendathu2k [1] OK to merge?
> 
> --
> Reply to this email directly, view it on GitHub [2], or unsubscribe
> [3].
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> 
> Links:
> ------
> [1] https://github.com/klendathu2k
> [2] 
> https://github.com/star-bnl/star-sw/pull/308#issuecomment-1227384733
> [3] 
> https://github.com/notifications/unsubscribe-auth/ANL4LVETY6AOKUFYRXADTEDV26DRBANCNFSM5OV3DOHQ
